### ISSUE TITLE
Styling question

### ISSUE THREAD

**Issue by joshcough** [thomashoneyman/purescript-halogen-realworld#46]
One more question for you guys. 

I started to really go through this last night because it's the most complete example and I'm wanting to use it, but the one thing I think is missing is talking anything about styling. I see you are using `demo.productionready.io/main.css`, and that's more or less opaque. Do you have any insights/links/tutorials/libraries explaining how to do this better? How to make Halogen sites look nice? 

The one thing I get with `purescript-elmish`, which I'm using on my project here `https://github.com/joshcough/chatbots/tree/master/frontend`, is that I can easily drop into react, import react-bootstrap, and just go. So I can actually start to make things look nice right away. Of course, `purescript-elmish` is missing a lot of things that Halogen has, which is why I'm looking here in the first place.

I'm just sorta looking for anything to help me tie it all together, if that makes sense. 

Thanks for putting this project together though, having all the comments in the code is really amazing.

**Comment by thomashoneyman** [comment]
Good question, @joshcough.

I'll answer below, but you may also want to participate in this Discourse thread: [CSS strategies for PureScript apps](https://discourse.purescript.org/t/css-strategies-for-purescript-apps).

On to Halogen-specific CSS strategies!

## External Stylesheets
Most folks using Halogen are simply using external stylesheets. It's what SlamData does, and we do the same thing at Awake Security (though we're on purescript-react), both using BEM-style naming. My previous company, CitizenNet, uses [functional CSS](https://tailwindcss.com) for [their component library](https://github.com/citizennet/purescript-ocelot), but it's still an external stylesheet. And in this project, Conduit, you see them again.

## Typed CSS
There are some attempts to support typed CSS for working with things like Bootstrap -- see, for example, the [purescript-css](https://github.com/slamdata/purescript-css) library. However, purescript-css has a reputation for being tedious to use, and afaik [SlamData just uses external stylesheets these days](https://discourse.purescript.org/t/css-strategies-for-purescript-apps/374/2?u=thomashoneyman).

There are several projects which implement the CSS from CSS libraries (but not the full components), linked below along with a mention to the author in case they'd like to weigh in:

* [purescript-halogen-bootstrap](https://github.com/slamdata/purescript-halogen-bootstrap), by SlamData 
* [purescript-halogen-bootstrap4](https://github.com/mschristiansen/purescript-halogen-bootstrap4) by @mchristiansen
* [purescript-pure-css](https://github.com/athanclark/purescript-pure-css) by @athanclark
* [purescript-halogen-bulma](https://github.com/Unisay/purescript-halogen-bulma) by @unisay

## UI Libraries
As far as Halogen goes, I'm not aware of any completed implementations of existing UI libraries like Bootstrap, Material UI, etc. which includes fully-implemented components. 

However: React Basic is a common alternative to Halogen for folks who want to use PureScript but want or need interop with React. You might be interested in @dwhitney's [project to implement Material UI](https://github.com/dwhitney/purescript-react-basic-mui) using React Basic. Most of the same ideas from this app apply to writing React Basic apps as well.

There are also open-source component libraries which can be used as inspiration, but which aren't generic enough to just drop in to your project, including those by [CitizenNet](https://github.com/citizennet/purescript-ocelot) (Halogen) and [Lumi](https://github.com/lumihq/purescript-lumi-components) (React Basic).

## CSS-in-JS / Inline CSS / Styled Components
@paulyoung has a few projects which support CSS-in-JS, including:

* Generate stylesheets (or strings) from CSS written in PureScript: [example](https://github.com/paulyoung/purescript-style/blob/9643b39b86fee771777e45d22f323433bad8ead0/test/Main.purs)
* Styled components for writing CSS directly in Halogen components: [example](https://github.com/paulyoung/purescript-styled-components/blob/83d0e4f34c33375ebb78c1e0d32e34e734a7e1ff/examples/buttons/src/Example/Button.purs)
* A system with utilities and more support for writing styled components: [example](https://github.com/paulyoung/purescript-styled-system/blob/8cbdd35967bd2e12b00d028ca191837396f3b3e9/examples/purescript-styled-components/src/Example/Box.purs)

There is an alternative approach by @danieljharvey named [purescript-stylesheet](https://github.com/danieljharvey/purescript-stylesheet), which updates the CSS object model directly and is used in libraries like [purescript-react-stylesheet](https://github.com/danieljharvey/purescript-react-stylesheet) and could be used in Halogen also.

I'm intrigued by this approach, and I know that styled-components is popular in JavaScript. I haven't tried the `purescript-stylesheet` stuff but we experimented with Paul Young's libraries briefly at CitizenNet and @davezuch found that there were some performance issues which prevented us from adopting it at the time. I'm not sure whether that's still the case.

---

In summary, I and most other folks I know use external stylesheets. Some go the extra mile and type the CSS with purescript-css. 

And there are other approaches -- typed CSS for Bootstrap and Bulma, an in-progress Material UI implementation for React Basic, and exploratory approaches for inline CSS, styled components, generating stylesheets, and so on, but I'm not sure any of them are ready for production without you needing to contribute some work as well.

**Comment by thomashoneyman** [comment]
Closing for inactivity, but feel free to continue commenting later.

**Comment by Benjmhart** [comment]
here's a quick look at my approach:

In terms of a css-in-purescript solution for Halogen, I’ve found the following is a pretty nice system for getting your styles set up:

mount a secondary tree in the head which contains your styles - it only needs a single component, which creates a Style tag.
for each component module, have a style module which exports the following:
the HTML classnames used by a module as either Strings or Halogen classnames
a concatenated set of css rules for your component (you can build these using the combinators and rules provided by the purescript-css package)
import and use the htrml classnames in your Halogen component - this mostly provides typesafety around your classes, that they are strings exported from the styles module rather than random strings that your app might not care about
in the Style component mounted in step 1. import the styles from each of your Halogen componets, concatenate them, and render them as text content of your style tag.
This is more or less what styled components and emotion do under the hood, and appears to be a low-overhead approach.

Unfortunately this approach isn’t great if you want to do style code-splitting.

in those cases you would probably need to mount secondary Style container tags when you get your second bundle.

**Comment by srghma** [comment]
hi @Benjmhart 

could you send an example of this approach

**Comment by srghma** [comment]
Also a question: how to eliminate dead code / check if classes were used in code

**Comment by srghma** [comment]
I found that css-modules work with purs-loader

https://imgur.com/a/QF11icB

https://github.com/srghma/purescript-halogen-material-components-web/blob/70a0d0b88b95fe3e8bc43a101e399c81c3a48ce0/docs-src/Pages/Buttons/CSS.module.css#L1

https://github.com/srghma/purescript-halogen-material-components-web/blob/70a0d0b88b95fe3e8bc43a101e399c81c3a48ce0/webpack.config.js#L55-L73
